llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -27469,6 +27469,21 @@ SDValue AArch64TargetLowering::LowerReductionToSVE(unsigned Opcode,
     VecOp = convertToScalableVector(DAG, ContainerVT, VecOp);
   }
 
+  // Lower VECREDUCE_ADD of nxv2i1-nxv16i1 to CNTP rather than UADDV.
+  if (ScalarOp.getOpcode() == ISD::VECREDUCE_ADD &&
+      VecOp.getOpcode() == ISD::ZERO_EXTEND) {
+    SDValue Vec = VecOp.getOperand(0);
+    EVT VecVT = Vec.getValueType();
+    if (VecVT.getVectorElementType() == MVT::i1) {
```

Comment:
There seems to be a check missing that the vector type is a legal predicate type.
What happens with this patch, if you'd start with a `<vscale x 32 x i1>` ?

Reply:
I thought that all types have already been legalised and therefore any scalable i1 vector is a legal predicate type.
Starting with a `<vscale x 32 x i1>` results in it getting split into two `nxv16i1`s which are then zero extended and added together before being add reduced. I don't match this so it gets lowered to a `UADDV_PRED`.

Reply:
Ah of course, I didn't realise this was in a Lower* function, rather than a DAG combine.

After splitting <vscale x 32 x i1>, I guess it could also use two `cntp` instructions (and add the results together). I guess this is more relevant for the extend from `<vscale x 16 x i1> -> <vscale x 16 x i64>` which generates a lot of punpklo/hi instructions, but it all depends on what code the LoopVectorizer generates.